### DISCUSSION TITLE
How to do logging with latest frb (v1.75.3 now)?

### DISCUSSION THREAD

**Discussion by dbsxdbsx** [fzyzcjy/flutter_rust_bridge#1220] [Q&A]
According to `_VisionUtilsRsImplExtended` from [reference](https://github.com/fzyzcjy/flutter_rust_bridge/issues/486#issuecomment-1147270588),  last year I made my own one  like this:
```dart
import 'package:project_name/gened_util_api.web.dart';
import 'package:flutter_rust_bridge/flutter_rust_bridge.dart';

import 'gened_util_api.dart';
import 'dart:ffi' as ffi;

class UtilApiImplExtend extends UtilApiImpl with FlutterRustBridgeSetupMixin {
  factory UtilApiImplExtend(ffi.DynamicLibrary dylib) =>
      UtilApiImplExtend._raw(UtilApiWire(dylib));

  UtilApiImplExtend._raw(super.inner) : super.raw() {
    setupMixinConstructor();
  }

  @override
  Future<void> setup() async {
    await rustSetUp(hint: FlutterRustBridgeSetupMixin.kHintSetup);
  }
}
```
And corresponding rust API `utlis_api.rs` like this:
```rust
pub fn rust_set_up() {
    crate::init_logger("./logs/");
}

// this API is used ONLY to trigger `rust_set_up()` from dart side
pub fn activate() {}
```
And use it in`main.dart` with: `utilApi.activate();` with `ffi.dart` like this:
```dart
final utilApi = UtilApiImplExtend(io.Platform.isIOS || io.Platform.isMacOS
    ? DynamicLibrary.executable()
    : DynamicLibrary.open(_dylib));
```
IIRC, this code is workable at that moment(November, 2022).   But now this code is full of syntax errors from the dart side.  I wonder what is the correct way to trigger log initialization now.

**Comment by fzyzcjy** [comment]
> full of syntax errors from the dart side

it would be great to show the errors (paste it, and/or add a screenshot if the red underlines are easier to read). Then I can see what is happening :)

**Reply by dbsxdbsx** [reply]
![image](https://github.com/fzyzcjy/flutter_rust_bridge/assets/17372655/2db1f737-ab1b-46f4-b597-da4291e22e53)
if 3 missing methods are added by the dart lint, then code becomes this:
![image](https://github.com/fzyzcjy/flutter_rust_bridge/assets/17372655/d5b4a16f-61d0-40e7-afeb-eecfc812d8bb)
But I don't know how to fill in the 3 methods.

I also read [frb-logging tutorial ](https://cjycode.com/flutter_rust_bridge/feature/logging.html?highlight=log#logging). But I guess there is no need to use `StreamSink` for just triggering logging initialization in rust for only logging in rust code with code like `log::info!(), log::warn!()...`.

**Reply by fzyzcjy** [reply]
Hmm `FlutterRustBridgeSetupMixin` becomes red - it should be mixed into your api

**Reply by fzyzcjy** [reply]
So, one way is to look at differences between the two versions (of support library as well as generated code). If it is still to vague, maybe binary search at the frb version, such that you can see which exact version causes the problem. I deliberately make each version only have minimal change, so this will be not hard.

**Reply by dbsxdbsx** [reply]
> So, one way is to look at differences between the two versions (of support library as well as generated code). If it is still to vague, maybe binary search at the frb version, such that you can see which exact version causes the problem. I deliberately make each version only have minimal change, so this will be not hard.

If a search is inevitable, I would do it. But before that, I wonder if you still use such kind of manually defined template for `crate::init_logger("./logs/");` this kind of API in your project. If so, could you provide the latest template please?

And personally, I hope there is no need to manually write the complex boilerplate only for revoking an API that has a global effect during the whole APP lifetime but is only triggered once.

**Reply by fzyzcjy** [reply]
I have not upgraded to the latest frb version indeed (since the API of my internal Rust codebase is already quite stable, and does not require new frb features), thus my code about logging has not been changed. However, I am going to migrate to Flutter 3.10 / Dart 3.0, and I do expect I need to upgrade frb in my project due to that. After that, I will tell you my updated logging solution!

> And personally, I hope there is no need to manually write the complex boilerplate only for revoking an API that has a global effect during the whole APP lifetime but is only triggered once.

Well, IMHO it is just to override `setup()` and the rest are boilerplate. If you find out a simpler solution feel free to share!
